### PR TITLE
Generate mask for STOMP in batched mode

### PR DESCRIPTION
**Pull request by jrecuerda**
The mask matrix used by STOM(self join) to discard trivial matches is
now generated in batched mode.

Unit tests have been added with the aim of covering all the corner cases
of the new implementation.

The new implementation allows the execution of STOMP for longer time
series which crashed with the previous one.

The performance impact measured by benchmarking shows that the new
implementation reduce the whole STOMP(self join) execution time by a
50%.

Some code has been reordered in src/khiva/matrix.cpp in order to avoid
polluting the global namespace with internal functions.

Make sure you have checked _all_ steps below.

### Description
- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Benchmarks
- [ ] My PR adds the following micro benchmarks __OR__ does not need benchmarks for this extremely good reason:

### Commits
- [ ] My commits have been squashed if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

## License
- [ ] Add a [Mozilla Public License 2.0 license header](http://mozilla.org/MPL/2.0/) to all the new files.

### Documentation
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.

### CODE REVIEW

File: benchmarks/matrixBench.cpp [hunk 1]
```diff
@@ -7,6 +7,8 @@
 #include <benchmark/benchmark.h>
 #include <khiva/matrix.h>
 #include <math.h>
+
```
Comment (dcuestam) [inline benchmarks/matrixBench.cpp]: Delete this blank line please

File: src/khiva/matrix.cpp [hunk 2]
```diff
@@ -6,238 +6,20 @@
 
 #include <khiva/matrix.h>
 #include <khiva/normalization.h>
+
```
Comment (dcuestam) [inline src/khiva/matrix.cpp]: I am not sure why is here this blank line? Do you want to differenciate between the header of khiva and others? If that is the case, don't remove it.

File: src/khiva/matrix.cpp [hunk 3]
```diff
@@ -746,12 +499,244 @@ void findBestN(af::array profile, af::array index, long m, long n, af::array &di
...
+    // The 3rd dimension reflects the number of reference time series
+    // The 4th dimension reflects the batch size of query subsequences from the query time series
+    // Reordering to match the mask band matrix dimensions
+    dist = af::reorder(dist, 3, 0, 1, 2);
+
+    // Increasing the distance using the mask band matrix to filter trivial matches
+    dist += 1 / EPSILON * mask.as(qt.type());
+
+    // The 1st dimension reflects the number of subsequences of the reference time series.
+    // The 2nd dimension reflects the number of reference time series.
```
Comment (dcuestam) [inline src/khiva/matrix.cpp]: Please be consistent with the comments, if you are going to add a point at the end of the comment, add it to every one, if not, write all the comments without it.

File: src/khiva/matrix.cpp [hunk 4]
```diff
@@ -746,12 +499,244 @@ void findBestN(af::array profile, af::array index, long m, long n, af::array &di
...
+    // the first dimension)
+    af::array sum_q2 = af::sum(af::pow(q, 2), 0);
+
+    // Calculate the distance and index profiles for all the combinations of query sequences and reference time series
+    calculateDistanceProfile(qt, a, sum_q, sum_q2, mean_t, sigma_t, mask, distance, index);
+}
+
+void mass(af::array q, af::array t, af::array a, af::array mean_t, af::array sigma_t, af::array &distance,
+          af::array &index) {
+    // Normalizing the query sequence. q can contain query sequences from multiple series
```
Comment (dcuestam) [inline src/khiva/matrix.cpp]: This code is the same as the above function, couldn't one of the reuse the other one?
